edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/ServiceIdentityTree.cs

```diff
@@ -0,0 +1,326 @@
+                .Where(s => s.Identity.IsModule && s.Identity.DeviceId == device.DeviceId)
+                .ToList()
+                .AsEnumerable());
+
+            foreach (ServiceIdentityTreeNode child in danglingChildren)
+            {
+                newNode.AddChild(child);
+            }
+        }
+
+        void RemoveSingleNode(string id)
+        {
+            if (this.nodes.TryGetValue(id, out ServiceIdentityTreeNode target))
+            {
+                // Unhook from parent
+                target.Parent.ForEach(p => p.RemoveChild(target));
+
+                // Unhook the children
+                target.RemoveAllChildren();
+
+                // Remove the node itself
+                this.nodes.Remove(id);
+            }
+        }
+
+        Option<ServiceIdentityTreeNode> FindDeviceByScopeId(string scopeId)
+        {
+            Preconditions.CheckNonWhiteSpace(scopeId, nameof(scopeId));
+
+            // Look for Edge devices with a matching scope ID
+            List<ServiceIdentityTreeNode> devices =
+                this.nodes
+                .Select(kvp => kvp.Value)
+                .Where(serviceIdentity => serviceIdentity.Identity.IsEdgeDevice && serviceIdentity.Identity.DeviceScope.Contains(scopeId))
+                .ToList();
+
+            if (devices.Count() > 0)
+            {
+                return Option.Some(devices.First());
+            }
+
+            return Option.None<ServiceIdentityTreeNode>();
+        }
+
+        /// <summary>
+        /// This is an object to wrap ServiceIdentity so that it can
+        /// point to a parent and act as a node in a tree.
+        /// </summary>
+        internal class ServiceIdentityTreeNode
+        {
+            // Only allow up to 5 Edge devices to be linked together
+            static readonly int MaxNestingDepth = 4;
+
+            public ServiceIdentity Identity { get; }
+            public Option<string> AuthChain { get; private set; }
+            public Option<ServiceIdentityTreeNode> Parent { get; private set; }
+
+            List<ServiceIdentityTreeNode> children;
+            int currentDepth;
+
+            public ServiceIdentityTreeNode(ServiceIdentity identity, Option<string> authChain)
+            {
+                this.Identity = Preconditions.CheckNotNull(identity);
+                this.Parent = Option.None<ServiceIdentityTreeNode>();
+                this.children = new List<ServiceIdentityTreeNode>();
+                this.AuthChain = authChain;
+                this.currentDepth = 0;
+            }
+
+            public void AddChild(ServiceIdentityTreeNode childNode)
+            {
+                if (!this.Identity.IsEdgeDevice)
+                {
+                    throw new ArgumentException($"{this.Identity.Id} is not an Edge device, only Edge devices can have children");
+                }
+
+                if (this.currentDepth >= MaxNestingDepth && childNode.Identity.IsEdgeDevice)
```

Comment:
I think MaxNestingDepth should be set to 5 and here should add 1 to currentDepth because it starts at 0